From 02426b5382a1392744c3db3ec93d79a7472e61fc Mon Sep 17 00:00:00 2001
From: Michael W <baddaemon87@gmail.com>
Date: Sun, 24 Feb 2019 21:27:51 +0100
Subject: [PATCH 3/3] Trust: Onboarding: Listen for locale changes

* When SuW is not yet done, the notification is already posted
* This results in an english notification text when the SuW is finished
  because the notification doesn't update when the locale changes
-> Listen for locale changes and post the notification again (= update)

Change-Id: I920a52c5c85c91adb7333a20d410a5464e80a812
---
 .../internal/TrustInterfaceService.java       | 30 +++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/lineage/lib/main/java/org/lineageos/platform/internal/TrustInterfaceService.java b/lineage/lib/main/java/org/lineageos/platform/internal/TrustInterfaceService.java
index 2dcb938..fb80c41 100644
--- a/lineage/lib/main/java/org/lineageos/platform/internal/TrustInterfaceService.java
+++ b/lineage/lib/main/java/org/lineageos/platform/internal/TrustInterfaceService.java
@@ -21,9 +21,11 @@ import android.app.NotificationChannel;
 import android.app.NotificationManager;
 import android.app.PendingIntent;
 import android.app.admin.DevicePolicyManager;
+import android.content.BroadcastReceiver;
 import android.content.ContentResolver;
 import android.content.Context;
 import android.content.Intent;
+import android.content.IntentFilter;
 import android.net.Uri;
 import android.os.Build;
 import android.os.IBinder;
@@ -32,6 +34,7 @@ import android.os.SELinux;
 import android.os.ServiceManager;
 import android.os.SystemProperties;
 import android.os.UserHandle;
+import android.provider.Settings;
 import android.util.Log;
 import android.util.Pair;
 import android.text.TextUtils;
@@ -64,6 +67,8 @@ public class TrustInterfaceService extends LineageSystemService {
     private static final String CHANNEL_NAME = "TrustInterface";
     private static final int ONBOARDING_NOTIFCATION_ID = 89;
 
+    private static final String SETTINGS_SECURE_USER_SETUP_COMPLETE = "user_setup_complete";
+
     private Context mContext;
     private NotificationManager mNotificationManager = null;
 
@@ -98,6 +103,9 @@ public class TrustInterfaceService extends LineageSystemService {
         // Onboard
         if (!hasOnboardedUser()) {
             postOnBoardingNotification();
+            if (!isUserSetupComplete()) {
+                registerLocaleChangedReceiver();
+            }
             return;
         }
 
@@ -354,6 +362,28 @@ public class TrustInterfaceService extends LineageSystemService {
                 LineageSettings.System.TRUST_INTERFACE_HINTED, 0) == 1;
     }
 
+    private boolean isUserSetupComplete() {
+        return Settings.Secure.getInt(mContext.getContentResolver(),
+                SETTINGS_SECURE_USER_SETUP_COMPLETE, 0) == 1;
+    }
+
+    private void registerLocaleChangedReceiver() {
+        IntentFilter filter = new IntentFilter(Intent.ACTION_LOCALE_CHANGED);
+        mContext.registerReceiver(mReceiver, filter);
+    }
+
+    private final BroadcastReceiver mReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            if (intent.getAction() == Intent.ACTION_LOCALE_CHANGED) {
+                postOnBoardingNotification();
+                if (isUserSetupComplete()) {
+                    mContext.unregisterReceiver(mReceiver);
+                }
+            }
+        }
+    };
+
     /* Service */
 
     private final IBinder mService = new ITrustInterface.Stub() {
-- 
2.17.1

